### PR TITLE
feat_: sending transaction improvements

### PR DESCRIPTION
**Pull request by saledjenic**
Instead of the complicated approach we used before, these changes should simplify sending transaction/s suggested by the Router. Many calculations that were done on the client side before will be fully handled on the statusgo side. These changes make the sending process the same for accounts (key pairs) migrated to a keycard and those stored locally in local keystore files.

Deprecated endpoints:
- `CreateMultiTransaction`
- `ProceedWithTransactionsSignatures`

Deprecated signal:
- `wallet.sign.transactions`

New endpoints:
- `BuildTransactionsFromRoute`
- `SendRouterTransactionsWithSignatures`

The flow for sending the best router suggested by the router:
- call `BuildTransactionsFromRoute`
- wait for the `wallet.router.sign-transactions` signal
- sign received hashes using `SignMessage` call or sign on keycard
- call `SendRouterTransactionsWithSignatures` with the signatures of signed hashes from the previous step
- `wallet.router.transactions-sent` signal will be sent after sending transactions or an error occurs

New signals:
- `wallet.router.sending-transactions-started` // notifies client that the sending transactions process started
- `wallet.router.sign-transactions` // notifies client about the list of transactions that need to be signed
- `wallet.router.transactions-sent` // notifies client about transactions that are sent
- `wallet.transaction.status-changed` // notifies about status of sent transactions

Tests for this PR will be added via this issue:
- https://github.com/status-im/status-go/issues/5831

### CODE REVIEW

File: services/wallet/api.go
Comment (dlipicar) [inline services/wallet/api.go]: we can make this API a bit more break-proof by using a struct (`parameters` or something) instead of plainly `slippagePercentage`

File: transactions/types.go
Comment (dlipicar) [inline transactions/types.go:70]: hmm not sure why we decided to just copy all the fields from `ethapi.TransactionArgs` (raw parameters needed to send a tx to the chain), I kind of like to keep that as a separate struct and then have all the remaining "metadata" as additional fields. Then the lower level modules that are not interested in the metadata can just consume the original type.

File: transactions/types.go
Comment (dlipicar) [inline transactions/types.go]: it's not clear what this parameter's meaning is without looking at the code, could you add some comment?

File: services/wallet/api.go
Comment (saledjenic) [reply services/wallet/api.go]: @dlipicar agree, good idea, done.

File: transactions/types.go
Comment (saledjenic) [reply transactions/types.go]: @dlipicar comment added.

File: transactions/types.go
Comment (saledjenic) [reply transactions/types.go:70]: @dlipicar this is the first step of improvements and simplifications I want to do, we can discuss that for some of the next steps for sure, but this was the easiest now since any bigger change will require updating `Transactor` that I don't want to touch yet, since I didn't want to break any endpoint that we already have and is being used by the mobile app.

File: transactions/types.go
Comment (alaibe) [inline transactions/types.go:23]: I guess this version will go away when deprecated code will be handle, maybe add a comment to remove this Version attribute?

Comment (saledjenic) [reply transactions/types.go:23]: Sure, will add. And yes, once the mobile app switches to the new tx sending flow, and we make sure that there is no part of the app that uses old sendargs, we can simplify `Transactor` a lot and remove the version property.
Once we move all TXs sendings to this flow, we will have a uniform flow across the app that is easier to maintain and extend.

**Review by alaibe — APPROVED** [review]

**Review by igor-sirotin — CHANGES_REQUESTED** [review]
@saledjenic I would like to open a discussion here and bring some attention.

-----

This PR is too big.
And there're 2 issues behind such huge PRs. Even Github files diff barely loads 🫠

1. Reviewers waste too much time and energy.

    And if they don't it means they didn't pay enough attention.
    If I was to review such PR carefully, it would take 1-2 hours of high concentration, just to keep all the details in mind. 

2. The chances to spot a bug are very low.
    
    Rarely can a developer keep so much stuff in mind. And even more rare they can assume how such code would break. 
    This basically makes the reviews useless.
    
    I admire you as a dev, the amount of stuff you can keep in mind is great. But other developers are not like you.

3. With due respect, there's no way even you can think of all the edge cases

    We're all humans and we do mistakes. And there're 2 ways to fight this: code review and testing. Code review is useless in such case. And there're no tests either.

4. There are no tests

    You know, me and you, we love this topic :heart: :smile: 
    
    I see that you opened a separate ticket. But why we can't implement it in a same PR?
    We have no rush now. There's no milestone. There's no deadline. The focus is quality and reliability. Isn't it?

-----

Also, such PRs take too much time to merge. And meanwhile you start preparing new changes on top of these. And then the PR is pushed to be reviewed and tested ASAP, as it's been open for so long, and there's another portion of code waiting. 

This process is unhealthy for the product and everyone involved.
It would go much faster and simpler for everyone, if the iterations were more granular.

-----

Having this said,
Would you please consider dividing this PR into a few smaller ones?

**Comment by saledjenic** [comment]
@igor-sirotin I appreciate your perspective and what you brought up, and I agree with most of them, but some things I see a bit differently. :)

First of all, I don't know why you are for smaller PRs instead of having multiple logically split commits in a single PR?
Some downsides that I see:
- if for example, I want to apply a change, like this one, in case more PRs are created I will need to rebase many times until they get reviewed
- if I don't do PRs one on top of another, but wait for it to be merged, before continuing, then I see that as wasting time and losing the context in terms of changes that I want to do
- while I'm in my 4th, 5th PR and I see that I forgot to add something that logically belongs to the 1st PR, a lot of work will be needed to add the change there and then rebase all PRs to get the change

> Would you please consider dividing this PR into a few smaller ones?

I can organize them in several commits instead, not PRs, hope that works?

> This PR is too big.

The size of the PR depends on the change that is done.  So if I want to rename a single function that is used in many places, I will have many changes and the PR will be big, but that change cannot be split. This may be a bad example, but what I want to say is I added a new function `BuildTransactionV2` to **the interface that is implemented by ~10 processors** (~10 different files and the implementation of that function is the same as the previous version that had to keep because of compatibility, just because the input params are different) **and that bloated a code**. So we have to look at the gist of changes. The gist is in `services/wallet/transfer/transaction_manager_route.go` all other changes are side effects. But that file cannot work without other changes.

> There are no tests

Correct, there are no tests here, yet, at the time of this PR, the mechanism for handling signals for functional tests was not implemented (not sure if that's merged even today), but because of that, I informed Anton D about that and created a new issue for adding tests for this part https://github.com/status-im/status-go/issues/5831 that was mentioned in the description of this PR.

**Comment by igor-sirotin** [comment]
@saledjenic thank you for being open for discussion ❤️ 

## Large PR split by commits VS. several small PRs

1. > The size of the PR depends on the change that is done

    Of course, a function rename or updating anything in a `vendor` folder might change 100s of files. But that's an obvious granular change. This PR doesn't seem to be one of these: 
    image
    
3. Of course, sometimes large PR is unavoidable, because changes are too tightly coupled. But:
    - This PR doesn't seem to be one (though I didn't dig into the changes)
    - It's better to follow a different process then: discuss and agree on the changes before making them, so that by the PR open moment everybody involved is already familiar with changes.
    - This is _usually_ a sign of that code needs to be refactored to decouple some parts

2. If changes can be decoupled, it's still better to have 5 small PRs in compare to 5 commits in a single PR.

    Comments to a single commit in a large PR will hold other 4 commits from being merged. 
  
4. Some comments:
    
    > in case more PRs are created I will need to rebase many times
    
    Yes, for you it will be more work. 
    But for the reviewers (which is more than one) it will be less. Overall less time will be spent. We should respect each other's time.
    
    I don't feel the problem of rebasing though. Is it that big of a thing today?
    
    > if I don't do PRs one on top of another, but wait for it to be merged, before continuing, then I see that as wasting time 
    
    Smaller PR's will be reviewed much faster.
    
    > while I'm in my 4th, 5th PR and I see that I forgot to add something that logically belongs to the 1st PR, a lot of work will be needed to add the change there
    
    Again, smaller PR's wil lbe reviewed much faster. You won't need to work on 4-5 PRs of the same context in parallel.
    
    But yes, again, it will be more time for you. But less time for everyone else.

## Tests

Soo, we actually have a policy in the repo:

https://github.com/status-im/status-go/blob/f04a9a8726b3e24686aacbc6f1fb00438e917670/_docs/policies/tests.md?plain=1#L8

It's totally fine if @antdanchenko implements the tests. But they should be introduced in the same PR, because we agreed on the intention to keep `develop` branch stable and release-ready.

> not sure if that's merged even today

FYI, it [was merged](https://github.com/status-im/status-go/pull/5806) a week ago 👌 

-----

Really, we're discussing things here that are a well-known good practices 🤔

**Comment by saledjenic** [comment]
@igor-sirotin nothing of that can be said in general, depends a lot from case to case. It depends on the scope ofc, but I am more for clean, logical and clear commits than many PRs, but again, it depends on the context, for changes done here, I cannot imagine having them in multiple PRs.

> But for the reviewers (which is more than one) it will be less. Overall less time will be spent. We should respect each other's time.

Regardless of the size of the PR reviewers know to filter out boilerplate and unavoidable following changes from the gist of the change which is the most important thing. Filter commits and use search if reviewing online, or the best, checkout the branch and review locally.

> It's totally fine if @antdanchenko implements the tests. But they should be introduced in the same PR, because we agreed on the intention to keep develop branch stable and release-ready.

Waiting for tests to be implemented in the same PR may take a while and that blocks new features. 

@igor-sirotin could you review this PR? 
@others please, review?

Also this is not final, more changes will be on this new sending tx flow.

**Comment by antdanchenko** [comment]
a bit late with my response, as I had several sick days last week.

@saledjenic, @igor-sirotin I've started with tests based on this PR, should be ready in a day or two, please let me know if the tests should be added within this PR or separately in develop

**Review by shivekkhurana — CHANGES_REQUESTED** [review]
Large PRs are not healthy. Given recent experiences, I don't feel comfortable merging changes without testing them for regressions on clients. This PR should be broken down into testable chunks.

It's not a refactor. It touches every critical wallet flow, and assumes the clients will adapt to it in future.

More context here:
https://discord.com/channels/1210237582470807632/1232982322874159104/1287745337150144522

**Comment by saledjenic** [comment]
@shivekkhurana this PR doesn't bring any breaking changes.
It also doesn't affect the mobile app at all.

It's also tested in the desktop app, as you can see here https://github.com/status-im/status-desktop/pull/16330

So please review it and let me know if there are any real changes that you request or not?

**Comment by saledjenic** [comment]
> @saledjenic, @igor-sirotin I've started with tests based on this PR, should be ready in a day or two, please let me know if the tests should be added within this PR or separately in develop

@antdanchenko thank you, the best would be to add those tests here in this PR as a new commit.

File: services/wallet/common/helpers.go
Comment (dlipicar) [inline services/wallet/common/helpers.go:13]: It might be silly since this function is straightforward, but I normally like writing a test for these kinds of things using real blockchain data. Basically get data for some approval from etherscan   and make sure the same input gives you the same output.

File: services/wallet/common/helpers.go
Comment (dlipicar) [inline services/wallet/common/helpers.go:26]: again, silly straightforward case, but a test would help set expectations on what the input string needs to look like

File: transactions/types.go
Comment (dlipicar) [inline transactions/types.go]: Hmm isn't this what the `To` field contains whenever we're not doing a simple ETH transfer?

File: services/wallet/common/helpers.go
Comment (saledjenic) [reply services/wallet/common/helpers.go:13]: @dlipicar test added, thanks.

File: services/wallet/common/helpers.go
Comment (saledjenic) [reply services/wallet/common/helpers.go:26]: @dlipicar test added, thanks.

File: transactions/types.go
Comment (saledjenic) [reply transactions/types.go]: @dlipicar well, this can be a topic for discussion as well. :)

My point is that we should be as explicit as possible, not think about what is `To` for example a recipient or a contract address, and that's why I added that new field and also plan to update it in one of the later steps the `Transactor` implementation (once mobile and desktop fully moved to that). 

When a client sends a tx from address A to address B, then A and B should be the sender and the recipient respectively, and any other address (contract) that should be used as an intermediary step to achieve that transfer should be set as a contract address field. In that case, the dev would be able to pack/send transactions easily without following the entire flow to figure out is "To" final destination, recipient or contract address.

What we have now, whenever I see the Transaction part I am always confused is the `To` a recipient address or a contract address and never know if packed the data correctly or not without following entire flow.

File: services/wallet/responses/router_transactions.go
Comment (dlipicar) [inline services/wallet/responses/router_transactions.go:46]: shouldn't we have a From/To amount here as well?

File: transactions/types.go
Comment (dlipicar) [reply transactions/types.go]: ahh ok, I think I got confused by the reuse of the old fields.

So, in `SendTxArgsVersion0` `To` has the old meaning, in `SendTxArgsVersion1` `To` is the destination address for a transfer. Is that correct? Same for `Value` I guess (only used for ETH transfers in `SendTxArgsVersion0`, used for any transaction in `SendTxArgsVersion1`).
Already mentioned it before, I know it's planned for later, but I'd really like to keep "raw blockchain" fields and "input/interpreted" fields separate...

File: services/wallet/api.go
Comment (dlipicar) [inline services/wallet/api.go:793]: I'm not super happy about status-go storing all this "temporary" data (then we need to introduce lifetime management and be super careful about when to clear that so we don't reach some inconsistent state), but as an improvement to make it less error-prone we could make sure all these temporarily cached things are tied to a specific UUID so we can be fully certain whatever we retrieve from the cache belongs to a particular Send/Swap/Bridge operation. No use case for now, but that would even probably let us lift the restriction of "only one route/transaction flow at a time".

File: services/wallet/responses/router_transactions.go
Comment (saledjenic) [reply services/wallet/responses/router_transactions.go:46]: If there is a need we can add it later, but for now we don't need it.
There is `RouterSentTransactions` type that is returned.
```
type RouterSentTransactions struct {
	SendDetails      *SendDetails             `json:"sendDetails"`
	SentTransactions []*RouterSentTransaction `json:"sentTransactions"`
}
```
so it contains send details where we already return "from" and "to" amounts and every tx from the `SendTransactions` list contains an amount that was used in that tx. So far only swap can have "to" amount and since every response contains send details + sent transactions, it's easy to deduce if an amount in tx was used as "from" or "to".

File: services/wallet/api.go
Comment (saledjenic) [reply services/wallet/api.go:793]: Well, that's the price of having a code organization like we have.

The sending Tx flow has to be done in a few steps (calculate the route, build tx, sign tx, send tx) there are 2 approaches:
- to return everything from each call and ask for all those params in the next step of the flow
- to keep states on the statusgo side and just let the client call the next step with certain uuid

Since the statusgo code is shared between desktop and mobile app, the first approach may result in what we have now, different client-side handlings for achieving the same result. That approach requires more lines of code than the second approach for sure. In the second approach, all that the client needs to do is to call functions in a certain order with uuid used when it was generating the best route, that's why I think the second approach will force clients to use the flow in the same way and will be easier for future maintaining. 

About sending flow data management on the statusgo side... route's data are cleared in the following cases:
- either on error when generating the best route (handled by statusgo)
- or asking for a new route (handled by statusgo)
- or explicitly by the client calling `StopSuggestedRoutesAsyncCalculation` (calling this func multiple times is fine)
- or when proceeding with sending (handled by statusgo)

clearing transactions' data is fully managed by the statusgo, data are cleared:
- either on error while building the Tx from the route
- or on error while sending the Tx
- or when the Tx is successfully sent

All that minimizes the burden on the client side.

**Comment by saledjenic** [comment]
@igor-sirotin @briansztamfater @dlipicar @shivekkhurana @alexjba please have a look and let me know if you're ok to merge this one? Or any other concerns?

@antdanchenko added tests for this PR, the last commit.

There is also the desktop PR that was tested and approved, here https://github.com/status-im/status-desktop/pull/16330, waiting for this to be merged.

This PR doesn't affect any current/old flows, thus the mobile app is not affected by this change.

File: integration-tests/docker-compose.test.status-go.yml
Comment (igor-sirotin) [inline integration-tests/docker-compose.test.status-go.yml]: leftover?

File: integration-tests/tests/test_cases.py
Comment (igor-sirotin) [inline integration-tests/tests/test_cases.py:136]: @antdanchenko I think this should be a bit different. 

If in the 0.5 seconds a few of same signals received, you'll randomly get one of them, instead of getting the first one received. So this introduces a flakiness right away.

Also, so far there's need to store all `received_signals`. In most (if not 100%) of cases you'll only need to wait for 1 specific signal, first received. 

-----

This doesn't need to be changed in this PR, but just raising your awareness.

File: integration-tests/tests/test_router.py
Comment (igor-sirotin) [inline integration-tests/tests/test_router.py:45]: Perhaps it makes sense to `verify_is_valid_json_rpc_response` inside `rpc_request`?

File: integration-tests/tests/test_router.py
Comment (igor-sirotin) [inline integration-tests/tests/test_router.py]:
```suggestion
        routes = self.wait_for_signal("wallet.suggested.routes")
        assert routes['event']['Uuid'] == _uuid
```

File: services/wallet/common/const.go
Comment (igor-sirotin) [inline services/wallet/common/const.go]: This is intended to be a `const`, but not possible with such initialisers. But having this as public `var` allows other packages to change the value.

In such cases in Go it's conventional to do like this:
```go
func ZeroAddress() ethCommon.Address {
    return ethCommon.Address{}
}

func ZeroBigIntValue() big.Int {
    return big.NewInt(0)
}
```

File: transactions/types.go
Comment (igor-sirotin) [inline transactions/types.go]: It's important to specify `json:...` tags explicitly.
Otherwise it's too easy to rename the field in Go not thinking of the API changes.

File: services/wallet/router/pathprocessor/processor_bridge_celar.go
Comment (igor-sirotin) [inline services/wallet/router/pathprocessor/processor_bridge_celar.go:305]: Maybe mark as `deprecated` then?

File: services/wallet/router/pathprocessor/processor_bridge_celar.go
Comment (igor-sirotin) [inline services/wallet/router/pathprocessor/processor_bridge_celar.go:368]: So all 4 `sendOrBuildV2` implementations are [not covered by tests](https://app.codecov.io/github/status-im/status-go/commit/da522e111b20d5320f4420b187df79870c163800/tree/services/wallet/router/pathprocessor). 

I understand there limitations, as anvil is not yet setup for bridging. 
But how can we be sure of the functionality state then? 🙁

File: services/wallet/api.go
Comment (igor-sirotin) [inline services/wallet/api.go]: I just saw this today in the `wallet/api.go`. Not a big deal, just sharing info
It should be `// Deprecated: ...` as per https://go.dev/wiki/Deprecated

File: services/wallet/api.go
Comment (igor-sirotin) [inline services/wallet/api.go]:
```suggestion
		// notify client that sending transactions started (has 3 steps, building txs, signing txs, sending txs)
```

File: services/wallet/api.go
Comment (igor-sirotin) [inline services/wallet/api.go:841]: Why do API endpoints end up so long? 🤔 
I thought that in most cases they should be a simple wrapper around some Go method

File: services/wallet/router/routes/router_path.go
Comment (igor-sirotin) [inline services/wallet/router/routes/router_path.go:58]: I think it would be good to unit-test such function

File: services/wallet/transfer/transaction_manager.go
Comment (igor-sirotin) [inline services/wallet/transfer/transaction_manager.go:57]: I'm not sure if it's used in API, but if it is, then please add explicit `json:"..."` tags

File: services/wallet/transfer/transaction_manager_route.go
Comment (igor-sirotin) [inline services/wallet/transfer/transaction_manager_route.go]: image

This is indeed too many arguments. Can you please combine them in a single/few struct?

File: services/wallet/transfer/transaction_manager_route.go
Comment (igor-sirotin) [inline services/wallet/transfer/transaction_manager_route.go]: Same here for number of arguments
image

**Comment by igor-sirotin** [comment]
Hmm, not sure why it shows `requested changes`
image

I only left comments 🤔 
image

File: services/wallet/common/const.go
Comment (saledjenic) [reply services/wallet/common/const.go]: Good point, updated, thanks.

File: transactions/types.go
Comment (saledjenic) [reply transactions/types.go]: Old params of `SendTxArgs` type were exposed to a client, and they all have `json:...`, but newly added params (added in this PR) are not, that's the difference, we've changed the flow. Those params are used only on the statusgo side and they are not exposed, also once mobile switches to this flow I guess we can remove json tags from old params as well.

File: services/wallet/router/pathprocessor/processor_bridge_celar.go
Comment (saledjenic) [reply services/wallet/router/pathprocessor/processor_bridge_celar.go:368]: Actually what matter is `BuildTransactionV2` call, which is a replacement for `BuildTransaction` and it's implemented by all processors, some processors use `sendOrBuildV2`, called by `BuildTransactionV2` some not. The only `BuildTransactionV2` that is currently covered by tests is `BuildTransactionV2` implemented by the `TransferProcessor`. @antdanchenko did that just to unblock this PR, and will add more tests later to cover other flows.

File: services/wallet/router/pathprocessor/processor_bridge_celar.go
Comment (saledjenic) [reply services/wallet/router/pathprocessor/processor_bridge_celar.go:305]: This is just TODO for devs, not sure if we need to mark it deprecated if it's private to package. Public API functions are already marked as deprecated.

File: services/wallet/api.go
Comment (saledjenic) [reply services/wallet/api.go]: Updated for endpoints marked as deprecated in this PR.

File: services/wallet/api.go
Comment (saledjenic) [reply services/wallet/api.go:841]: Cause moving all that out of API requires many other changes. I agree, we should do that through another PR.

File: integration-tests/tests/test_cases.py
Comment (antdanchenko) [reply integration-tests/tests/test_cases.py:136]: this method just checks (every 0.5) if there is an expected signal in the received_signals list, 

the list itself is being appended by every signal that matches self.await_signals here https://github.com/status-im/status-go/pull/5807/files#diff-396e333ecb0055730c471697ae8e74841d20f7308a2f1f9ad03fd49851518bf6R123

File: integration-tests/tests/test_router.py
Comment (antdanchenko) [reply integration-tests/tests/test_router.py:45]: we have negative tests as well, as result this two are separated

File: services/wallet/router/routes/router_path.go
Comment (saledjenic) [reply services/wallet/router/routes/router_path.go:58]: Added, thanks for reminding me. ;)

File: services/wallet/transfer/transaction_manager.go
Comment (saledjenic) [reply services/wallet/transfer/transaction_manager.go:57]: No, it's not, it's just for internal use.

File: services/wallet/transfer/transaction_manager_route.go
Comment (saledjenic) [reply services/wallet/transfer/transaction_manager_route.go]: Updated.

File: services/wallet/transfer/transaction_manager_route.go
Comment (saledjenic) [reply services/wallet/transfer/transaction_manager_route.go]: Updated.

**Comment by saledjenic** [comment]
@igor-sirotin all your comments are addressed, thanks for the review.

**Comment by churik** [comment]
@saledjenic as just today we merged all swaps related PRs + all fixes after last wallet refactoring, we'll retest it briefly one more time, thank you!

**Comment by VolodLytvynenko** [comment]
Hey @saledjenic , thanks for the PR! I’ve checked to see if it affects mobile, and I can confirm it doesn’t while also fixing the ERC-20 approval [issues](https://github.com/status-im/status-mobile/issues/21332) on Arbitrum. Please let me know if any last commits are added, and I'll quickly recheck to ensure it still doesn't impact mobile. I promise it won't take long :)

**Comment by saledjenic** [comment]
Thanks for checking @VolodLytvynenko.

Nothing new was added, since I just rebase this PR from time to time, cause it's opened for a long time already and I have other PRs based on it and that's why you see/receive email about changes.

File: transactions/types.go
Comment (igor-sirotin) [reply transactions/types.go]: @saledjenic If so, then we should explicitly mark the fields to be ignored in json with `json:"-"`

Comment (saledjenic) [reply transactions/types.go]: @igor-sirotin updated, thanks.

**Review by igor-sirotin — APPROVED** [review]
The code's good, thanks @saledjenic for the fixes.

----

I still see a lot of uncovered new code, which goes against the agreed policy.
I won't hold this PR, it's been around for 1 month already. But I'm worried about this, so please let's discuss this and agree on our processes for the new PRs.

**Pull request merged**
